src/Blammo/Logging/Terminal.hs

```diff
@@ -21,57 +21,82 @@ module Blammo.Logging.Terminal
 import Control.Monad.Logger.Aeson
 import Data.Aeson
 import Data.Aeson.Compat (KeyMap)
 import qualified Data.Aeson.Compat as Key
 import qualified Data.Aeson.Compat as KeyMap
 import Data.ByteString (ByteString)
 import qualified Data.ByteString.Lazy as BSL
+import Data.List (sortOn)
 import Data.Maybe (fromMaybe)
 import Data.Text (Text, pack)
 import qualified Data.Text as T
-import Data.Text.Encoding (encodeUtf8)
 import Data.Time (defaultTimeLocale, formatTime)
 import qualified Data.Vector as V
 
-reformatTerminal :: Bool -> LogLevel -> ByteString -> ByteString
-reformatTerminal useColor logLevel bytes = fromMaybe bytes $ do
+reformatTerminal :: Int -> Bool -> LogLevel -> ByteString -> ByteString
+reformatTerminal breakpoint useColor logLevel bytes = fromMaybe bytes $ do
   LoggedMessage {..} <- decode $ BSL.fromStrict bytes
 
   let
     colors@Colors {..} = getColors useColor
 
-    logTimestampText =
-      dim $ pack $ formatTime defaultTimeLocale "%F %X" loggedMessageTimestamp
+    logTimestampPiece = logPiece dim $ pack $ formatTime
+      defaultTimeLocale
+      "%F %X"
+      loggedMessageTimestamp
 
-    logLevelText = case logLevel of
-      LevelDebug -> gray $ padTo 9 "debug"
-      LevelInfo -> green $ padTo 9 "info"
-      LevelWarn -> yellow $ padTo 9 "warn"
-      LevelError -> red $ padTo 9 "error"
-      LevelOther x -> blue $ padTo 9 x
+    logLevelPiece = case logLevel of
+      LevelDebug -> logPiece gray $ padTo 9 "debug"
+      LevelInfo -> logPiece green $ padTo 9 "info"
+      LevelWarn -> logPiece yellow $ padTo 9 "warn"
+      LevelError -> logPiece red $ padTo 9 "error"
+      LevelOther x -> logPiece blue $ padTo 9 x
 
     loggedSourceAsMap =
       foldMap (KeyMap.singleton "source" . String) loggedMessageLogSource
 
-  pure $ encodeUtf8 $ mconcat
-    [ logTimestampText <> " "
-    , "[" <> logLevelText <> "] "
-    , bold $ padTo 31 loggedMessageText
-    , colorizeKeyMap colors loggedSourceAsMap
-    , colorizeKeyMap colors loggedMessageThreadContext
-    , colorizeKeyMap colors loggedMessageMeta
-    ]
-
-colorizeKeyMap :: Colors -> KeyMap Value -> Text
-colorizeKeyMap Colors {..} km
-  | KeyMap.null km = ""
-  | otherwise = " " <> T.intercalate " " keyValues
- where
-  keyValues = map (uncurry renderPair) $ KeyMap.toList km
+    logPrefixPiece =
+      mconcat [logTimestampPiece <> " ", "[" <> logLevelPiece <> "] "]
+
+    logMessagePiece = logPiece bold $ padTo 31 loggedMessageText
+
+    logAttrsPiece = mconcat
+      [ colorizeKeyMap " " colors loggedSourceAsMap
+      , colorizeKeyMap " " colors loggedMessageThreadContext
+      , colorizeKeyMap " " colors loggedMessageMeta
+      ]
+
+    oneLineLogPiece = mconcat [logPrefixPiece, logMessagePiece, logAttrsPiece]
 
-  renderPair k v = cyan (Key.toText k) <> "=" <> magenta (fromValue v)
+    multiLineLogPiece =
+      let
+        shift = "\n" <> LogPiece.offset (LogPiece.visibleLength logPrefixPiece)
```

Comment:
Will `"\n"` work in windows? Is that something we care to support.

Reply:
Good point. I have no idea if any of this will work on windows, but if we code for it without much downside I'm into it.

Is there a library with a cross-platform `newline` function? Like we try to use `(</>)` as much as we can.

Reply:
I think I'm going to leave this for now. We'll cross the windows bridge when/if we get an Issue, since I assume the reporter can help us with the right strategies.

Reply:
Sounds good. I'm very afraid now: https://stackoverflow.com/a/5944062